src/ralph/lib/transitions/async.py

```diff
@@ -0,0 +1,132 @@
+"""
+Asynchronous runner for transitions
+"""
+import logging
+
+from django.db import transaction
+
+from ralph.lib.external_services.models import JobStatus
+from ralph.lib.transitions.models import (
+    _check_instances_for_transition,
+    _check_action_with_instances,
+    _order_actions_by_requirements,
+    _prepare_action_data,
+    TransitionJob,
+    TransitionJobAction,
+    TransitionJobActionStatus
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RescheduleAsyncTransitionActionLater(Exception):
+    pass
+
+
+class AsyncTransitionError(Exception):
+    pass
+
+
+class MoreThanOneStartedActionError(AsyncTransitionError):
+    pass
+
+
+class FailedActionError(AsyncTransitionError):
+    pass
+
+
+def _check_previous_actions(job, executed_actions):
+    """
+    Check if:
+    * none of previously executed actions has failed
+    * there is max 1 started action
+    """
+    started_actions = 0
+    for action in executed_actions:
+        if action.status == TransitionJobActionStatus.FAILED:
+            logger.warning(
+                'Action {} has failed. Marked job {} as failed too'.format(
+                    action, job
+                )
+            )
+            job.status = JobStatus.FAILED
+            job.save()
+            raise FailedActionError()
+        elif action.status == TransitionJobActionStatus.STARTED:
+            started_actions += 1
+
+    if started_actions > 1:
+        logger.error('Job {} has more than one started action'.format(
+            job
+        ))
+        job.fail()
+        raise MoreThanOneStartedActionError()
+
+
+def run_async_transition(job_id):
+    transition_job = TransitionJob.objects.get(pk=job_id)
+    transition = transition_job.transition
+    obj = transition_job.obj
+    _check_instances_for_transition([obj], transition)
+    _check_action_with_instances([obj], transition)
+
+    # check if this job isn't already finished
+    if not transition_job.is_running:
+        logger.warning(
+            'Runnig previously ended transition job: {}'.format(transition_job)
+        )
+        return
+    # make sure that none of previous actions has failed
+    executed_actions = list(transition_job.transition_job_actions.all())
+    try:
+        _check_previous_actions(transition_job, executed_actions)
+    except AsyncTransitionError:
+        return
+
+    completed_actions_names = set([
+        tja.action_name for tja in executed_actions
+        if tja.status != TransitionJobActionStatus.STARTED
+    ])
+    for action in _order_actions_by_requirements(transition.actions.all(), obj):
```

Comment:
TODO: make sure that first executed action if the one with STARTED status (if there is already exactly one with this status)